cfg/builder/builder_finalize.cc

```diff
@@ -386,32 +396,42 @@ void CFGBuilder::fillInBlockArguments(core::Context ctx, const CFG::ReadsAndWrit
             for (auto it = cfg.forwardsTopoSort.rbegin(); it != cfg.forwardsTopoSort.rend(); ++it) {
                 BasicBlock *bb = *it;
                 auto &upperBoundsForBlock = upperBounds2[bb->id];
-                int sz = upperBoundsForBlock.size();
+                const auto sz = upperBoundsForBlock.size();
                 for (BasicBlock *edge : bb->backEdges) {
                     if (edge != cfg.deadBlock()) {
-                        upperBoundsForBlock.insert(writesByBlock[edge->id].begin(), writesByBlock[edge->id].end());
-                        upperBoundsForBlock.insert(upperBounds2[edge->id].begin(), upperBounds2[edge->id].end());
+                        setUnionInplace(upperBoundsForBlock, writesByBlock[edge->id]);
+                        setUnionInplace(upperBoundsForBlock, upperBounds2[edge->id]);
                     }
                 }
-                changed = changed || (upperBounds2[bb->id].size() != sz);
+
+                changed = changed || sz != upperBoundsForBlock.size();
             }
         }
     }
     {
         Timer timeit(ctx.state.tracer(), "upperBoundsMerge");
         /** Combine two upper bounds */
         for (auto &it : cfg.basicBlocks) {
-            auto set2 = upperBounds2[it->id];
-
-            int set1Sz = set2.size();
-            int set2Sz = upperBounds1[it->id].size();
-            it->args.reserve(set1Sz > set2Sz ? set2Sz : set1Sz);
-            for (auto el : upperBounds1[it->id]) {
-                if (set2.find(el) != set2.end()) {
-                    it->args.emplace_back(el);
+            const auto &set1 = upperBounds1[it->id];
+            const auto &set2 = upperBounds2[it->id];
+            auto set1It = set1.begin();
+            auto set2It = set2.begin();
+            // Note: The loop enqueues arguments in sorted order. If args has any members already, we'll need to sort
+            // it!
+            ENFORCE_NO_TIMER(it->args.empty());
+            while (set1It != set1.end() && set2It != set2.end()) {
+                const auto set1El = *set1It;
+                const auto set2El = *set2It;
+                if (set1El == set2El) {
+                    it->args.emplace_back(set1El);
```

Comment:
Since `set1` and `set2` are sorted, and `it->args` is empty at the start of this loop, we know that `it->args` will be sorted too.

Reply:
This would be a good comment to move into the code.